Sources/OutlineView/OutlineViewDataSource.swift

```diff
@@ -43,4 +89,151 @@ where Data.Element: Identifiable {
+              !pasteboardItems.isEmpty
+        else { return nil }
+        
+        let decodedItems = pasteboardItems.compactMap {
+            dropReceiver.readPasteboard(item: $0)
+        }
+        
+        let dropTarget = DropTarget<Data.Element>(
+            items: decodedItems,
+            intoElement: item.map(typedItem)?.value,
+            childIndex: childIndex == NSOutlineViewDropOnItemIndex ? nil : childIndex,
+            isItemExpanded: { item in
+                let typed = OutlineViewItem<Data>(value: item, children: self.childrenSource)
+                return outlineView.isItemExpanded(typed)
+            }
+        )
+        return dropTarget
+    }
+    
+    func outlineView(
+        _ outlineView: NSOutlineView,
+        validateDrop info: NSDraggingInfo,
+        proposedItem item: Any?,
+        proposedChildIndex index: Int
+    ) -> NSDragOperation {
+        
+        guard let dropTarget = dropTargetData(in: outlineView, dragInfo: info, item: item, childIndex: index)
+        else { return [] }
+        
+        let validationResult = dropReceiver.validateDrop(target: dropTarget)
+
+        switch validationResult {
+            
+        case .copy:
+            return .copy
+        case .move:
+            return .move
+        case .deny:
+            return []
+        case let .copyRedirect(item, childIndex):
+            let redirectTarget = item.map { OutlineViewItem<Data>(value: $0, children: childrenSource) }
+            outlineView.setDropItem(redirectTarget, dropChildIndex: childIndex ?? NSOutlineViewDropOnItemIndex)
+            return .copy
+        case let .moveRedirect(item, childIndex):
+            let redirectTarget = item.map { OutlineViewItem<Data>(value: $0, children: childrenSource) }
+            outlineView.setDropItem(redirectTarget, dropChildIndex: childIndex ?? NSOutlineViewDropOnItemIndex)
+            return .move
+        }
+        
+    }
+
+    func outlineView(
+        _ outlineView: NSOutlineView,
+        acceptDrop info: NSDraggingInfo,
+        item: Any?,
+        childIndex index: Int
+    ) -> Bool {
+        
+        guard let dropTarget = dropTargetData(in: outlineView, dragInfo: info, item: item, childIndex: index)
+        else { return false }
+        
+        // Perform `acceptDrop(target:)` to get result of drop
+        let res = dropReceiver.acceptDrop(target: dropTarget)
+        
+        // If drop was successful, but onto an unexpanded item, a quirk of NSOutlineView
+        // will cause the item to expand with the newly dropped item already in it, which
+        // messes up OutlineViewUpdater's attempt to update the contents of the item (by adding
+        // the dragged items onto items already displayed in the parent item).
+        // To fix this, we briefly listen for itemDidExpandNotification, and reload the item just
+        // after it expands, then cancel our listener.
```

Comment:
Not sure I fully understand this _hack_. Is there a reason why this can't be done deterministically?

Is there a chance two drops in rapid succession (may be drops queued up by the system) mess up this workaround be canceling the notification for the first item?

Either way, if there is no way to make this deterministic, it is imperative that the comment mentions any theoretical downsides of the notification not triggering or triggering because of an unrelated expansion.

Reply:
@ZofiaBernadetta  this one was very tricky to track down, so I'm inclined to say it's not really possible to do deterministically. My solution is definitely a hack, but I wasn't able to find any cases of rapid-succession drops like you mention. I think those would mess it up, I'm just not sure how to test for it :-) 

Here's a recording of the bug in action, where the hack isn't in place. As you can see, when the contents of the folder that opens due to a successful drop is doubled-- this happens because returning `true` from `outlineView(_:acceptDrop:item:childIndex:)` causes the item to expand, and at the same time the change of the data source causes `OutlineViewUpdater` to fire, and that in turn adds rows during the `for change in diff {` loop.

I could keep trying to find another way around this, but it may end up being just as hacky. I'll give myself some time to think it over and report back :-) 

https://user-images.githubusercontent.com/1621465/211034253-0767833c-dee4-4391-837b-762dddfb6665.mov

Reply:
Thank you @RCCoop for clearly describing the issue. I definitely understand it better now. I can't think of a solution off the top of my head. Non deterministic hack like this that affect out understanding of the model often have a way of creating hard to debug bugs.

I will think about this a little more. If I can't come up with something in the next couple of days, I don't think this should prevent you from merging your PR.

Reply:
Glad it's not a dealbreaker :-) I think the real killer here is that any calls to NSOutlineView to find out what contents it has end up going to the developer-provided data source, so it seems like we're kind of stuck there.

Reply:
Spent some more time looking into this. I believe the real problem here is that the `TreeMap` isn't updated by the time `OutlineViewUpdater.performUpdates` gets called.

This happens because the "item will expand" notification handler method gets called out of sequence due to it being scheduled asynchronously on to the main queue.

If you remove `.receive(on: DispatchQueue.main)` from your subscriptions in the init of `OutlineViewDataSource`, the drag and drop will start to work as expected without this hack.

https://github.com/RCCoop/OutlineView/blob/drag-n-drop/Sources/OutlineView/OutlineViewDataSource.swift#L41-L52

I would say that this is a general bug. The UI related outline view notifications are always published on the main queue.

Reply:
Wow! Good find! I thought I had tested it without the `receive(on:)` modifiers, but that must have been my imagination. I'm really glad you found that... makes things much cleaner. Fixed here: https://github.com/Sameesunkaria/OutlineView/pull/10/commits/2b7a5322dd47c62f7bb7f06dcadceeb2861f3e80